src/Language/DifferentialDatalog/Syntax.hs

```diff
@@ -702,17 +708,50 @@ data RuleRHS = RHSLiteral   {rhsPolarity :: Bool, rhsAtom :: Atom}
                    -- Group-by expression must be a tuple consisting of variable names, e.g., `group_by(q)`, `group_by(p,q,r)`.
                    rhsGroupBy :: Expr
                }
-             | RHSFlatMap   {rhsVars :: Expr, rhsMapExpr :: Expr}
-             | RHSInspect   {rhsInspectExpr :: Expr}
-             deriving (Eq, Ord)
+             | RHSFlatMap {
+                   rhsPos :: Pos,
+                   rhsVars :: Expr,
+                   rhsMapExpr :: Expr
+               }
+             | RHSInspect {
+                   rhsPos :: Pos,
+                   rhsInspectExpr :: Expr
+               }
+
+instance WithPos RuleRHS where
+    pos = rhsPos
+    atPos r p = r{rhsPos = p}
+
+instance Eq RuleRHS where
+    (==) (RHSLiteral _ p1 a1)     (RHSLiteral _ p2 a2)      = (p1, a1) == (p2, a2)
+    (==) (RHSCondition _ e1)      (RHSCondition _ e2)       = e1 == e2
+    (==) (RHSGroupBy _ v1 p1 g1)  (RHSGroupBy _ v2 p2 g2)   = (v1, p1, g1) == (v2, p2, g2)
+    (==) (RHSFlatMap _ v1 e1)     (RHSFlatMap _ v2 e2)      = (v1, e1) == (v2, e2)
+    (==) (RHSInspect _ e1)        (RHSInspect _ e2)         = e1 == e2
+    (==) _                        _                         = False
+
+rhsrank :: RuleRHS -> Int
+rhsrank RHSLiteral{}   = 0
+rhsrank RHSCondition{} = 1
+rhsrank RHSGroupBy{}   = 2
+rhsrank RHSFlatMap{}   = 3
+rhsrank RHSInspect{}   = 4
+
+instance Ord RuleRHS where
```

Comment:
forgot who needs Ord

Reply:
`Set` and `Map`.

Reply:
This is Ord for RuleRHS, not for the data.

Reply:
Any time we need e.g., a set of rules, we need `Ord Rule`, which in turn requires `Ord RuleRHS`.